// playwright-bdd/dist/run/bddData/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBddDataFromTest = exports.BddDataManager = void 0;
const createTestStep_1 = require("../../cucumber/createTestStep");
const utils_1 = require("../../utils");
const utils_2 = require("../../playwright/utils");
const BDD_DATA_ANNOTATION_NAME = '__bddData';
class BddDataManager {
    constructor(testInfo, testMeta, uri) {
        this.testInfo = testInfo;
        this.data = {
            uri,
            pickleLocation: testMeta.pickleLocation,
            steps: [],
        };
        this.save({ create: true });
    }
    registerStep(stepDefinition, stepText, pwStepLocation) {
        const step = (0, createTestStep_1.createTestStep)(stepDefinition, stepText);
        this.data.steps.push({
            pwStepLocation: (0, utils_1.stringifyLocation)(pwStepLocation),
            stepMatchArgumentsLists: step.stepMatchArgumentsLists || [],
        });
        this.save();
    }
    save({ create = false } = {}) {
        (0, utils_2.updateAnnotation)(this.testInfo, {
            type: BDD_DATA_ANNOTATION_NAME,
            description: JSON.stringify(this.data),
        }, { create });
    }
}
exports.BddDataManager = BddDataManager;
function getBddDataFromTest({ annotations }) {
    const annotationIndex = annotations.findIndex(isBddDataAnnotation);
    const annotation = annotations[annotationIndex];
    const bddData = annotation?.description
        ? JSON.parse(annotation.description)
        : undefined;
    return { bddData, annotationIndex };
}
exports.getBddDataFromTest = getBddDataFromTest;
function isBddDataAnnotation(annotation) {
    return annotation.type === BDD_DATA_ANNOTATION_NAME;
}
